src/inmanta/server/services/environmentservice.py

```diff
@@ -322,11 +383,49 @@ def get_depended_by(self) -> List[str]:
     @protocol.handle(methods_v2.environment_setting_delete, env="tid", key="id")
     async def environment_setting_delete(self, env: data.Environment, key: str) -> ReturnValue[None]:
         try:
+            original_env = env.to_dto()
             await env.unset(key)
             warnings = await self._setting_change(env, key)
             result: ReturnValue[None] = ReturnValue(response=None)
             if warnings:
                 result.add_warnings(warnings)
+            await self.notify_listeners(EnvironmentAction.updated, env.to_dto(), original_env)
             return result
         except KeyError:
             raise NotFound()
+
+    def register_listener_for_multiple_actions(self, listener, actions: Set[EnvironmentAction]) -> None:
+        """
+            Should only be called during pre-start
+        :param listener: The listener to register
+        :param actions: type of actions the listener is interested in
+        """
+        for action in actions:
+            self.register_listener(listener, action)
+
+    def register_listener(self, listener: EnvironmentListener, action: EnvironmentAction) -> None:
+        """
+            Should only be called during pre-start
+        :param listener: The listener to register
+        :param action: type of action the listener is interested in
+        """
+        self.listeners[action].append(listener)
+
+    def remove_listener(self, action: EnvironmentAction, listener: EnvironmentListener) -> None:
+        self.listeners[action].remove(listener)
+
+    async def notify_listeners(
+        self, action: EnvironmentAction, updated_env: model.Environment, original_env: Optional[model.Environment] = None
+    ) -> None:
+        for listener in self.listeners[action]:
+            try:
+                if action == EnvironmentAction.created:
+                    await listener.environment_action_created(updated_env)
+                if action == EnvironmentAction.deleted:
+                    await listener.environment_action_deleted(updated_env)
+                if action == EnvironmentAction.cleared:
+                    await listener.environment_action_cleared(updated_env)
+                if action == EnvironmentAction.updated:
+                    await listener.environment_action_updated(updated_env, original_env)
+            except Exception as e:
+                LOGGER.warning(f"Notifying listener of {action} failed with the following exception", e)
```

Comment:
The parameter `e` is not used in the warning message. This results in an exception (`TypeError: not all arguments converted during string formatting`). We should add a test case with a listener that throws an exception.